core/tx_list.go

```diff
@@ -552,12 +593,32 @@ func (l *txPricedList) Discard(slots int, force bool) (types.Transactions, bool)
 
 // Reheap forcibly rebuilds the heap based on the current remote transaction set.
 func (l *txPricedList) Reheap() {
-	reheap := make(priceHeap, 0, l.all.RemoteCount())
-
-	l.stales, l.remotes = 0, &reheap
+	start := time.Now()
+	l.stales = 0
+	l.urgent.list = make([]*types.Transaction, 0, l.all.RemoteCount())
 	l.all.Range(func(hash common.Hash, tx *types.Transaction, local bool) bool {
-		*l.remotes = append(*l.remotes, tx)
+		l.urgent.list = append(l.urgent.list, tx)
 		return true
 	}, false, true) // Only iterate remotes
-	heap.Init(l.remotes)
+	heap.Init(&l.urgent)
+
+	// balance out the two heaps by moving the worse half of transactions into the
+	// floating heap
+	// Note: Discard would also do this before the first eviction but Reheap can do
+	// is more efficiently. Also, Underpriced would work suboptimally the first time
+	// if the floating queue was empty.
+	floatingCount := len(l.urgent.list) * floatingRatio / (urgentRatio + floatingRatio)
```

Comment:
Shouldn't floatingCount be only the excess above the urgent cap? E.g. if we have 4+1K txs configured, we should only overflow if above 4K. Currently if I have 10txs, I'll overflow 2, which imho doesn't make sense.

Reply:
See below